Jakarta Data repositories can only be defined for default persistence unit

### Describe the bug

I read Hibernate ORM section describing the Jakarta Data support here https://quarkus.io/guides/hibernate-orm#jakarta-data and it says nothing about limitations as for with persistence unit can be used. I also checked https://hibernate.org/repositories/ and https://jakarta.ee/specifications/data/1.0/jakarta-data-1.0. So I went to https://quarkus.io/guides/hibernate-orm#persistence-unit-active and used configuration described there, something like this:

```
quarkus.datasource."pg".username=my-user
quarkus.datasource."pg".password=my-pwd
quarkus.datasource."pg".jdbc.url=this-is-set-dynamically
quarkus.datasource."pg".db-kind=postgresql
quarkus.hibernate-orm."pg".packages=io.quarkus.ts.jakarta.data.db
quarkus.hibernate-orm."pg".datasource=pg
```

and repository:

```
package io.quarkus.ts.jakarta.data.db;

import jakarta.data.repository.CrudRepository;
import jakarta.data.repository.Repository;

@Repository
public interface FruitCrudRepository extends CrudRepository<Fruit, Long> {
}
```

and entity:

```
package io.quarkus.ts.jakarta.data.db;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotBlank;

@Table(name = "fruit")
@Entity
public class Fruit {

    public Fruit() {
    }

    public Fruit(String name, long id) {
        this.name = name;
        this.id = id;
    }

    @Id
    Long id;

    @NotBlank
    String name;

    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }
}

```

and build fails.

### Expected behavior

I'd appreciate documented limitations, better exception or more hints what I am doing wrong. Thanks

### Actual behavior

```
[INFO] --- quarkus:999-SNAPSHOT:build (build) @ jakarta-data ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  10.941 s
[INFO] Finished at: 2025-07-09T17:52:28+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (build) on project jakarta-data: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: jakarta.enterprise.inject.spi.DeploymentException: jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.hibernate.StatelessSession and qualifiers [@Default]
[ERROR] 	- injection target: parameter 'session' of io.quarkus.ts.jakarta.data.db.FruitCrudRepository_ constructor
[ERROR] 	- declared on CLASS bean [types=[jakarta.data.repository.DataRepository<io.quarkus.ts.jakarta.data.db.Fruit, java.lang.Long>, jakarta.data.repository.CrudRepository<io.quarkus.ts.jakarta.data.db.Fruit, java.lang.Long>, io.quarkus.ts.jakarta.data.db.FruitCrudRepository_, io.quarkus.ts.jakarta.data.db.FruitCrudRepository, jakarta.data.repository.BasicRepository<io.quarkus.ts.jakarta.data.db.Fruit, java.lang.Long>, java.lang.Object], qualifiers=[@Default, @Any], target=io.quarkus.ts.jakarta.data.db.FruitCrudRepository_]
[ERROR] 	The following beans match by type, but none has matching qualifiers:
[ERROR] 	- SYNTHETIC bean [types=[org.hibernate.StatelessSession, java.lang.Object], qualifiers=[@Any, @io.quarkus.hibernate.orm.PersistenceUnit("pg"), @Named("pg")], target=n/a]
[ERROR] 
[ERROR] 
[ERROR] 	at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1594)
[ERROR] 	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:341)
[ERROR] 	at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:178)
[ERROR] 	at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:488)
[ERROR] 	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:889)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.hibernate.StatelessSession and qualifiers [@Default]
[ERROR] 	- injection target: parameter 'session' of io.quarkus.ts.jakarta.data.db.FruitCrudRepository_ constructor
[ERROR] 	- declared on CLASS bean [types=[jakarta.data.repository.DataRepository<io.quarkus.ts.jakarta.data.db.Fruit, java.lang.Long>, jakarta.data.repository.CrudRepository<io.quarkus.ts.jakarta.data.db.Fruit, java.lang.Long>, io.quarkus.ts.jakarta.data.db.FruitCrudRepository_, io.quarkus.ts.jakarta.data.db.FruitCrudRepository, jakarta.data.repository.BasicRepository<io.quarkus.ts.jakarta.data.db.Fruit, java.lang.Long>, java.lang.Object], qualifiers=[@Default, @Any], target=io.quarkus.ts.jakarta.data.db.FruitCrudRepository_]
[ERROR] 	The following beans match by type, but none has matching qualifiers:
[ERROR] 	- SYNTHETIC bean [types=[org.hibernate.StatelessSession, java.lang.Object], qualifiers=[@Any, @io.quarkus.hibernate.orm.PersistenceUnit("pg"), @Named("pg")], target=n/a]
[ERROR] 
[ERROR] 
[ERROR] 	at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:547)
[ERROR] 	at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:689)
[ERROR] 	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:326)
[ERROR] 	... 12 more
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

### How to Reproduce?

See the issue description. I can push what I have if you have any issues with reproducing this.

### Output of `uname -a` or `ver`

Fedora 42

### Output of `java -version`

Temurin 21

### Quarkus version or git rev

999-SNAPSHOT

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/docs/src/main/asciidoc/hibernate-orm.adoc
+++ b/docs/src/main/asciidoc/hibernate-orm.adoc
@@ -1796,10 +1796,10 @@ public interface MyRepository extends CrudRepository<MyEntity, Integer> { // <1>
 
 }
 ----
-1. To skip the boilerplate definition of CRUD operations,
+<1> To skip the boilerplate definition of CRUD operations,
 we can use one of the available interfaces (e.g. `CrudRepository` or `BasicRepository`).
-2. Adding custom queries with parameters is as easy as providing your query string to the `@Query` annotation.
-3. If the basic CRUD operations from the Jakarta Data interfaces are not enough,
+<2> Adding custom queries with parameters is as easy as providing your query string to the `@Query` annotation.
+<3> If the basic CRUD operations from the Jakarta Data interfaces are not enough,
 we can always add a custom one, in this case a delete operation that removes `MyEntity`s by name.
 
 And then the repository can be used as any other bean:
@@ -1817,10 +1817,28 @@ public class MyEntityResource {
         repository.insert(entity);
     }
 
-	// ...
+    // ...
 
 }
 ----
+
+[NOTE]
+====
+When working with non-default persistence units, remember to specify the persistence unit name the repository is targeting
+in the `dataStore` attribute of the repository annotation:
+
+[source,java]
+----
+@Repository(dataStore = "other") // <1>
+public interface MyNonDefaultPURepository {
+    // ...
+}
+----
+<1> Pass the name of the non-default persistence unit to the repository annotation (`other` in this example).
+
+Review the https://jakarta.ee/specifications/data/1.0/apidocs/jakarta.data/jakarta/data/repository/repository#dataStore()[Javadoc of the `@Repository` annotation]
+to learn more about the `dataStore` attribute.
+====
 
 Please refer to the corresponding https://hibernate.org/repositories/[Hibernate Data Repositories]
 and https://jakarta.ee/specifications/data/1.0/jakarta-data-1.0[Jakarta Data]

--- a/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/MyEntityResource.java
+++ b/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/MyEntityResource.java
@@ -1,0 +1,70 @@
+package io.quarkus.it.hibernate.processor.data;
+
+import java.util.List;
+
+import jakarta.data.Order;
+import jakarta.data.Sort;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import jakarta.transaction.Transactional;
+import jakarta.ws.rs.Consumes;
+import jakarta.ws.rs.DELETE;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.NotFoundException;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+
+import org.jboss.resteasy.reactive.RestPath;
+
+import io.quarkus.it.hibernate.processor.data.pudefault.MyEntity;
+import io.quarkus.it.hibernate.processor.data.pudefault.MyEntity_;
+import io.quarkus.it.hibernate.processor.data.pudefault.MyRepository;
+
+@ApplicationScoped
+@Produces("application/json")
+@Consumes("application/json")
+@Path("/data/")
+public class MyEntityResource {
+
+    @Inject
+    MyRepository repository;
+
+    @POST
+    @Transactional
+    public void create(MyEntity entity) {
+        repository.insert(entity);
+    }
+
+    @GET
+    public List<MyEntity> get() {
+        return repository.findAll(Order.by(Sort.asc(MyEntity_.NAME))).toList();
+    }
+
+    @GET
+    @Transactional
+    @Path("/by/name/{name}")
+    public MyEntity getByName(@RestPath String name) {
+        List<MyEntity> entities = repository.findByName(name);
+        if (entities.isEmpty()) {
+            throw new NotFoundException();
+        }
+        return entities.get(0);
+    }
+
+    @POST
+    @Transactional
+    @Path("/rename/{before}/to/{after}")
+    public void rename(@RestPath String before, @RestPath String after) {
+        MyEntity byName = getByName(before);
+        byName.name = after;
+        repository.update(byName);
+    }
+
+    @DELETE
+    @Transactional
+    @Path("/by/name/{name}")
+    public void deleteByName(@RestPath String name) {
+        repository.delete(name);
+    }
+}

--- a/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/MyOtherEntityResource.java
+++ b/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/MyOtherEntityResource.java
@@ -1,4 +1,4 @@
-package io.quarkus.it.hibernate.jpamodelgen.data;
+package io.quarkus.it.hibernate.processor.data;
 
 import java.util.List;
 
@@ -17,31 +17,35 @@ import jakarta.ws.rs.Produces;
 
 import org.jboss.resteasy.reactive.RestPath;
 
+import io.quarkus.it.hibernate.processor.data.puother.MyOtherEntity;
+import io.quarkus.it.hibernate.processor.data.puother.MyOtherEntity_;
+import io.quarkus.it.hibernate.processor.data.puother.MyOtherRepository;
+
 @ApplicationScoped
 @Produces("application/json")
 @Consumes("application/json")
-@Path("/data/")
-public class MyEntityResource {
+@Path("/data/other")
+public class MyOtherEntityResource {
 
     @Inject
-    MyRepository repository;
+    MyOtherRepository repository;
 
     @POST
     @Transactional
-    public void create(MyEntity entity) {
+    public void create(MyOtherEntity entity) {
         repository.insert(entity);
     }
 
     @GET
-    public List<MyEntity> get() {
-        return repository.findAll(Order.by(Sort.asc(MyEntity_.NAME))).toList();
+    public List<MyOtherEntity> get() {
+        return repository.findAll(Order.by(Sort.asc(MyOtherEntity_.NAME))).toList();
     }
 
     @GET
     @Transactional
     @Path("/by/name/{name}")
-    public MyEntity getByName(@RestPath String name) {
-        List<MyEntity> entities = repository.findByName(name);
+    public MyOtherEntity getByName(@RestPath String name) {
+        List<MyOtherEntity> entities = repository.findByName(name);
         if (entities.isEmpty()) {
             throw new NotFoundException();
         }
@@ -52,7 +56,7 @@ public class MyEntityResource {
     @Transactional
     @Path("/rename/{before}/to/{after}")
     public void rename(@RestPath String before, @RestPath String after) {
-        MyEntity byName = getByName(before);
+        MyOtherEntity byName = getByName(before);
         byName.name = after;
         repository.update(byName);
     }

--- a/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/pudefault/MyEntity.java
+++ b/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/pudefault/MyEntity.java
@@ -1,0 +1,30 @@
+package io.quarkus.it.hibernate.processor.data.pudefault;
+
+import jakarta.persistence.Column;
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.Id;
+
+@Entity
+public class MyEntity {
+
+    @Id
+    @GeneratedValue
+    public Integer id;
+
+    @Column(unique = true)
+    public String name;
+
+    MyEntity() {
+    }
+
+    public MyEntity(String name) {
+        this.name = name;
+    }
+
+    @Override
+    public String toString() {
+        return "MyOrmEntity [id=" + id + ", name=" + name + "]";
+    }
+
+}

--- a/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/pudefault/MyRepository.java
+++ b/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/pudefault/MyRepository.java
@@ -1,4 +1,4 @@
-package io.quarkus.it.hibernate.jpamodelgen.data;
+package io.quarkus.it.hibernate.processor.data.pudefault;
 
 import java.util.List;
 import java.util.stream.Stream;

--- a/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/puother/MyOtherEntity.java
+++ b/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/puother/MyOtherEntity.java
@@ -1,4 +1,4 @@
-package io.quarkus.it.hibernate.jpamodelgen.data;
+package io.quarkus.it.hibernate.processor.data.puother;
 
 import jakarta.persistence.Column;
 import jakarta.persistence.Entity;
@@ -6,7 +6,7 @@ import jakarta.persistence.GeneratedValue;
 import jakarta.persistence.Id;
 
 @Entity
-public class MyEntity {
+public class MyOtherEntity {
 
     @Id
     @GeneratedValue
@@ -15,10 +15,10 @@ public class MyEntity {
     @Column(unique = true)
     public String name;
 
-    MyEntity() {
+    MyOtherEntity() {
     }
 
-    public MyEntity(String name) {
+    public MyOtherEntity(String name) {
         this.name = name;
     }
 

--- a/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/puother/MyOtherRepository.java
+++ b/integration-tests/hibernate-orm-data/src/main/java/io/quarkus/it/hibernate/processor/data/puother/MyOtherRepository.java
@@ -1,0 +1,25 @@
+package io.quarkus.it.hibernate.processor.data.puother;
+
+import java.util.List;
+import java.util.stream.Stream;
+
+import jakarta.data.Order;
+import jakarta.data.repository.CrudRepository;
+import jakarta.data.repository.Delete;
+import jakarta.data.repository.Find;
+import jakarta.data.repository.Query;
+import jakarta.data.repository.Repository;
+
+@Repository(dataStore = "other")
+public interface MyOtherRepository extends CrudRepository<MyOtherEntity, Integer> {
+
+    @Find
+    Stream<MyOtherEntity> findAll(Order<MyOtherEntity> order);
+
+    @Query("select e from MyOtherEntity e where e.name like :name")
+    List<MyOtherEntity> findByName(String name);
+
+    @Delete
+    void delete(String name);
+
+}

--- a/integration-tests/hibernate-orm-data/src/main/resources/application.properties
+++ b/integration-tests/hibernate-orm-data/src/main/resources/application.properties
@@ -1,2 +1,9 @@
 quarkus.datasource.jdbc.max-size=8
-quarkus.hibernate-orm.database.generation=drop-and-create
+quarkus.hibernate-orm.packages=io.quarkus.it.hibernate.processor.data.pudefault
+quarkus.hibernate-orm.schema-management.strategy=drop-and-create
+quarkus.hibernate-orm.schema-management.create-schemas=true
+# Define non-default PU so that we can configure a custom JSON format mapper. The default PU is using the default mapper.
+quarkus.hibernate-orm."other".datasource=<default>
+quarkus.hibernate-orm."other".packages=io.quarkus.it.hibernate.processor.data.puother
+quarkus.hibernate-orm."other".schema-management.strategy=drop-and-create
+quarkus.hibernate-orm."other".schema-management.create-schemas=true

--- a/integration-tests/hibernate-orm-data/src/test/java/io/quarkus/it/hibernate/processor/data/HibernateOrmDataInGraalIT.java
+++ b/integration-tests/hibernate-orm-data/src/test/java/io/quarkus/it/hibernate/processor/data/HibernateOrmDataInGraalIT.java
@@ -1,4 +1,4 @@
-package io.quarkus.it.hibernate.jpamodelgen.data;
+package io.quarkus.it.hibernate.processor.data;
 
 import io.quarkus.test.junit.QuarkusIntegrationTest;
 

--- a/integration-tests/hibernate-orm-data/src/test/java/io/quarkus/it/hibernate/processor/data/HibernateOrmDataTest.java
+++ b/integration-tests/hibernate-orm-data/src/test/java/io/quarkus/it/hibernate/processor/data/HibernateOrmDataTest.java
@@ -1,48 +1,50 @@
-package io.quarkus.it.hibernate.jpamodelgen.data;
+package io.quarkus.it.hibernate.processor.data;
 
 import static io.restassured.RestAssured.given;
 import static org.hamcrest.CoreMatchers.containsString;
 import static org.hamcrest.CoreMatchers.equalTo;
 
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
 
+import io.quarkus.it.hibernate.processor.data.pudefault.MyEntity;
 import io.quarkus.test.junit.QuarkusTest;
 import io.restassured.http.ContentType;
 
 @QuarkusTest
 public class HibernateOrmDataTest {
-    private static final String ROOT = "/data";
 
-    @Test
-    public void staticMetamodel() {
+    @ValueSource(strings = { "/data", "/data/other" })
+    @ParameterizedTest
+    public void repositoryDefaultPu(String root) {
         // Create/retrieve
         given()
                 .pathParam("name", "foo")
                 .contentType(ContentType.JSON)
-                .when().get(ROOT + "/by/name/{name}")
+                .when().get(root + "/by/name/{name}")
                 .then()
                 .statusCode(404);
         given()
                 .contentType(ContentType.JSON)
-                .when().get(ROOT)
+                .when().get(root)
                 .then()
                 .statusCode(200)
                 .body(equalTo("[]"));
         given()
                 .body(new MyEntity("foo"))
                 .contentType(ContentType.JSON)
-                .when().post(ROOT)
+                .when().post(root)
                 .then()
                 .statusCode(204);
         given()
                 .pathParam("name", "foo")
                 .contentType(ContentType.JSON)
-                .when().get(ROOT + "/by/name/{name}")
+                .when().get(root + "/by/name/{name}")
                 .then()
                 .statusCode(200);
         given()
                 .contentType(ContentType.JSON)
-                .when().get(ROOT)
+                .when().get(root)
                 .then()
                 .statusCode(200)
                 .body(containsString("\"foo\""));
@@ -51,20 +53,20 @@ public class HibernateOrmDataTest {
         given()
                 .pathParam("name", "bar")
                 .contentType(ContentType.JSON)
-                .when().get(ROOT + "/by/name/{name}")
+                .when().get(root + "/by/name/{name}")
                 .then()
                 .statusCode(404);
         given()
                 .pathParam("before", "foo")
                 .pathParam("after", "bar")
                 .contentType(ContentType.JSON)
-                .when().post(ROOT + "/rename/{before}/to/{after}")
+                .when().post(root + "/rename/{before}/to/{after}")
                 .then()
                 .statusCode(204);
         given()
                 .pathParam("name", "bar")
                 .contentType(ContentType.JSON)
-                .when().get(ROOT + "/by/name/{name}")
+                .when().get(root + "/by/name/{name}")
                 .then()
                 .statusCode(200);
 
@@ -72,13 +74,13 @@ public class HibernateOrmDataTest {
         given()
                 .pathParam("name", "bar")
                 .contentType(ContentType.JSON)
-                .when().delete(ROOT + "/by/name/{name}")
+                .when().delete(root + "/by/name/{name}")
                 .then()
                 .statusCode(204);
         given()
                 .pathParam("name", "bar")
                 .contentType(ContentType.JSON)
-                .when().get(ROOT + "/by/name/{name}")
+                .when().get(root + "/by/name/{name}")
                 .then()
                 .statusCode(404);
     }

